pfcpiface/ip_pool.go

```diff
@@ -19,38 +19,50 @@ type IPPool struct {
 	inventory map[uint64]net.IP
 }
 
-func (i *IPPool) DeallocIP(seid uint64) error {
-	i.mu.Lock()
-	defer i.mu.Unlock()
+// NewIPPool creates a new pool of IP addresses with the given subnet.
+// The smallest supported size is a /30.
+func NewIPPool(poolSubnet string) (*IPPool, error) {
+	ip, ipnet, err := net.ParseCIDR(poolSubnet)
+	if err != nil {
+		return nil, err
+	}
 
-	ip, ok := i.inventory[seid]
-	if !ok {
-		log.Warnln("Attempt to dealloc non-existent session", seid)
-		return ErrInvalidArgumentWithReason("seid", seid, "can't dealloc non-existent session")
+	i := &IPPool{
+		inventory: make(map[uint64]net.IP),
 	}
 
-	delete(i.inventory, seid)
-	i.freePool = append(i.freePool, ip) // Simply append to enqueue.
-	log.Traceln("Deallocated session ", seid, "IP", ip)
+	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); inc(ip) {
+		ipVal := make(net.IP, len(ip))
+		copy(ipVal, ip)
+		i.freePool = append(i.freePool, ipVal)
+	}
 
-	return nil
+	if len(i.freePool) < 2 {
+		return nil, ErrInvalidArgumentWithReason("NewIPPool", poolSubnet, "pool subnet is too small to use as a pool")
+	}
+
+	// remove network address and broadcast address
+	i.freePool = i.freePool[1 : len(i.freePool)-1]
+
+	return i, nil
 }
 
 func (i *IPPool) LookupOrAllocIP(seid uint64) (net.IP, error) {
 	i.mu.Lock()
 	defer i.mu.Unlock()
 
-	if len(i.freePool) == 0 {
-		return nil, ErrOperationFailedWithReason("IP allocation", "ip pool empty")
-	}
-
 	// Try to find an exiting session and return the allocated IP.
 	ip, found := i.inventory[seid]
 	if found {
 		log.Traceln("Found existing session", seid, "IP", ip)
 		return ip, nil
 	}
 
+	// Check capacity before new allocations.
+	if len(i.freePool) == 0 {
+		return nil, ErrOperationFailedWithReason("IP allocation", "ip pool empty")
+	}
```

Comment:
Doing this earlier leads to a bug when to pool is exhausted and lookups are attempted.